File: stone_lib/src/main/java/com/github/klee0kai/stone/types/wrappers/Wrapper.java
package com.github.klee0kai.stone.types.wrappers;

public interface Wrapper {

    /**
     * Provide wrapped object.
     *
     * @param wrapperCl type of wrapper
     * @param original  original object
     * @param <Wr>      type of wrapper
     * @param <T>       type of providing original object
     * @return wrapped object provider
     */
    <Wr, T> Wr wrap(Class<Wr> wrapperCl, T original);

}

File: stone_multiplatform/src/commonMain/kotlin/com/github/klee0kai/stone/wrappers/creators/Wrapper.kt
package com.github.klee0kai.stone.wrappers.creators

import kotlin.reflect.KClass

interface Wrapper {
    /**
     * Provide wrapped object.
     *
     * @param wrapperCl type of wrapper
     * @param original  original object
     * @param <Wr>      type of wrapper
     * @param <T>       type of providing original object
     * @return wrapped object provider
    </T></Wr> */
    fun <Wr : Any, T : Any> wrap(wrapperCl: KClass<Wr>?, original: T?): Wr?
}
